File: backend/app/schemas/stock.py
"""
Pydantic schemas for StockTracker API request/response validation.

These schemas define the data contracts between frontend and backend,
ensuring all data is properly validated and typed at API boundaries.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


# ── Request Schemas ───────────────────────────────────────────────────

class StockCreate(BaseModel):
    """Schema for adding a new stock ticker."""
    symbol: str = Field(
        ...,
        min_length=1,
        max_length=10,
        description="Stock symbol code, e.g. '600519' for A-share or 'AAPL' for US",
        examples=["600519", "000001", "AAPL", "MSFT"],
    )
    market: str = Field(
        default="CN",
        pattern=r"^(CN|US)$",
        description="Market identifier: 'CN' for China A-share, 'US' for US stocks",
    )


class ConfigUpdate(BaseModel):
    """Schema for updating scheduler configuration at runtime."""
    refresh_interval_seconds: int | None = Field(
        None,
        ge=10,
        le=3600,
        description="Data refresh interval in seconds (min 10, max 3600)",
    )
    market_hours_only: bool | None = Field(
        None,
        description="Only refresh during each market's trading hours (CN/US independently)",
    )


# ── Response Schemas ──────────────────────────────────────────────────

class StockResponse(BaseModel):
    """Full stock data response — maps directly from the Stock ORM model."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    symbol: str
    name: str | None = None
    market: str = "CN"

    # Price
    current_price: float | None = None
    open_price: float | None = None
    high_price: float | None = None
    low_price: float | None = None
    close_price: float | None = None

    # Volume
    volume: int | None = None
    turnover: float | None = None
    turnover_rate: float | None = None

    # Change
    change_amount: float | None = None
    change_percent: float | None = None
    amplitude: float | None = None

    # Fundamentals
    market_cap: float | None = None
    circulating_market_cap: float | None = None
    pe_ratio: float | None = None
    pb_ratio: float | None = None

    # 52-week range
    high_52w: float | None = None
    low_52w: float | None = None

    # Timestamps
    last_trade_time: datetime | None = None
    updated_at: datetime | None = None
    created_at: datetime | None = None


class StockListResponse(BaseModel):
    """Wrapper for a list of stocks with metadata."""
    count: int = Field(description="Total number of tracked stocks")
    stocks: list[StockResponse] = Field(description="List of stock records")


class SchedulerStatusResponse(BaseModel):
    """Current scheduler status information."""
    running: bool
    refresh_interval_seconds: int
    market_hours_only: bool
    next_run: str | None = None


class MessageResponse(BaseModel):
    """Generic message response for simple operations."""
    message: str
    success: bool = True

File: frontend/src/types/stock.ts
/**
 * TypeScript type definitions for StockTracker.
 *
 * These types mirror the backend Pydantic schemas
 * and provide type safety across the frontend codebase.
 */

// ── Stock Data Types ─────────────────────────────────────────────────

/** Full stock record as returned by `GET /api/stocks` */
export interface Stock {
  id: number;
  symbol: string;
  name: string | null;
  market: "CN" | "US" | "HK";

  // Price
  current_price: number | null;
  open_price: number | null;
  high_price: number | null;
  low_price: number | null;
  close_price: number | null;

  // Volume
  volume: number | null;
  turnover: number | null;
  turnover_rate: number | null;

  // Change
  change_amount: number | null;
  change_percent: number | null;
  amplitude: number | null;

  // Fundamentals
  market_cap: number | null;
  circulating_market_cap: number | null;
  pe_ratio: number | null;
  pb_ratio: number | null;

  // 52-week range
  high_52w: number | null;
  low_52w: number | null;

  // Timestamps (ISO strings from backend)
  last_trade_time: string | null;
  updated_at: string | null;
  created_at: string | null;
}

/** Response wrapper for GET /api/stocks */
export interface StockListResponse {
  count: number;
  stocks: Stock[];
}

/** Request body for POST /api/stocks */
export interface StockCreateRequest {
  symbol: string;
  market: "CN" | "US" | "HK";
}

/** Supported market type */
export type MarketType = "CN" | "US" | "HK";

// ── Scheduler & Config Types ─────────────────────────────────────────

/** Real-time open/closed status for each market */
export interface MarketStatusInfo {
  cn_open: boolean;
  us_open: boolean;
  hk_open: boolean;
}

/** Response from GET /api/scheduler/status */
export interface SchedulerStatus {
  running: boolean;
  refresh_interval_seconds: number;
  market_hours_only: boolean;
  next_run: string | null;
  market_status: MarketStatusInfo;
}

/** Request body for PATCH /api/config */
export interface ConfigUpdateRequest {
  refresh_interval_seconds?: number;
  market_hours_only?: boolean;
}

/** Generic message response from the API */
export interface MessageResponse {
  message: string;
  success: boolean;
}

// ── UI State Types ───────────────────────────────────────────────────

/** Current loading / error state for the stocks hook */
export interface StocksState {
  stocks: Stock[];
  loading: boolean;
  error: string | null;
  lastRefresh: Date | null;
}

/** Sort direction for table columns */
export type SortDirection = "ascend" | "descend" | null;

/** Sort configuration */
export interface SortConfig {
  field: keyof Stock;
  direction: SortDirection;
}

// ── Market Index Types ───────────────────────────────────────────────

/** Real-time quote for a single market index */
export interface IndexQuote {
  symbol: string;
  name: string;
  market: "CN" | "US" | "HK";
  current_price: number | null;
  change_amount: number | null;
  change_percent: number | null;
}

/** Response from GET /api/indices */
export interface IndicesResponse {
  cn: IndexQuote[];
  hk: IndexQuote[];
  us: IndexQuote[];
}

// ── Index Minute Data Types ─────────────────────────────────────────

/** A single minute-level price point */
export interface IndexMinutePoint {
  time: string;
  price: number;
}

/** Intraday minute data for a single market index */
export interface IndexMinuteData {
  symbol: string;
  name: string;
  market: "CN" | "HK" | "US";
  date: string | null;
  prev_close: number | null;
  points: IndexMinutePoint[];
}

/** Response from GET /api/indices/minute */
export interface IndicesMinuteResponse {
  cn: IndexMinuteData[];
  hk: IndexMinuteData[];
  us: IndexMinuteData[];
}
